### ISSUE TITLE
openssl s_time does not work well

### ISSUE THREAD

**Issue by judowha** [openssl/openssl#17964]
I am testing the tls1.3 handshake speed using the s_time command. 
I created a server with an ECDSAP256 certificate and a server with an RSA3072 certificate. Both of them are signed by a self-signed root CA.
The test result shows here:
Server with RSA
```
7380 connections in 5.02s; 1470.12 connections/user sec, bytes read 0
7380 connections in 31 real seconds, 0 bytes read per connection
```
Server with P256
```
19755 connections in 16.49s; 1198.00 connections/user sec, bytes read 0
19755 connections in 31 real seconds, 0 bytes read per connection
```
In theory, ECDSA should perform better than RSA in the same security level but the test result shows the opposite.

**There is another problem.** I find that the test result for adding the `-new` flag and adding the `-reuse` flag is basically the same. 
Server with RSA with `-new`
```
6221 connections in 4.24s; 1467.22 connections/user sec, bytes read 0
6221 connections in 31 real seconds, 0 bytes read per connection
```
Server with RSA with `-reuse`
```
6325 connections in 4.26s; 1484.74 connections/user sec, bytes read 0
6325 connections in 31 real seconds, 0 bytes read per connection
```
Then I added `-msg` on the server-side. I found that **for both situations, the `CertificateVerify` message is sent.** However, in tls1.3 doing session resumption should not send the `CertificateVerify` message.

My test commands are here, `$1` will be replaced by folder names:
```
openssl s_server -key $1_test/server/server.key -cert $1_test/server/server.pem -accept 4330 -CAfile $1_test/rootCA/demoCA/root.pem -verify_return_error -msg -msgfile outfile.txt
```
```
openssl s_time -connect localhost:4330 -CAfile rsa_test/rootCA/demoCA/root.pem
```
My questions are:
**Why the test result using s_time is different from the theory?**
**What are the difference between using `-new` and `-reuse` flags?**

**Comment by baentsch** [comment]
Can you reproduce the issue in "stock builds", i.e., builds not done by you? Do the results remain as reported across different and longer repetitions? What is your execution environment? Dedicated machine (architecture? OS? etc) or VM (notoriously known for "surprising" handshake timing results)?

**Comment by judowha** [comment]
> Can you reproduce the issue in "stock builds", i.e., builds not done by you? Do the results remain as reported across different and longer repetitions? What is your execution environment? Dedicated machine (architecture? OS? etc) or VM (notoriously known for "surprising" handshake timing results)?

I have tried to reproduce the issue on different computer with different OpenSSL version (1.1.1m, 1.1.1, 1.1.1f) and I always get the same result. I will try to find my friend to test it again next. I also run the test for many times so the result should be reliable. My execution environment is Ubuntu 18.04 LTS, not virtual machine

**Comment by baentsch** [comment]
Thanks for the additional information. Could you possibly share the script(s) to reproduce your runs? Key/cert gen through server/client setup and actual `s_time` calls? Most importantly, I'm curious what value is used for `verify` [depth] (and/or how the results differ with different cert check depths) -- but also I'd like to run it in my test env...

**Comment by judowha** [comment]
> Thanks for the additional information. Could you possibly share the script(s) to reproduce your runs? Key/cert gen through server/client setup and actual `s_time` calls? Most importantly, I'm curious what value is used for `verify` [depth] (and/or how the results differ with different cert check depths) -- but also I'd like to run it in my test env...

Yeah sure!
My experiment files are here

[test1.zip](https://github.com/openssl/openssl/files/8376097/test1.zip)
This is the test describe in this post. It contains two folder named `ecc_test` and `rsa_test`. In `ecc_test` contains a root CA and a server. The server has a certificate signed by root CA using ECDSAP256. In `rsa_test` everything are the same except that the signature algorithm is RSA3072

**ALL `$1` in the command need to be replaced by folder name (i.e. `ecc_test` or `rsa_test`).**
The server is set up using this command:
`openssl s_server -key $1_test/server/server.key -cert $1_test/server/server.pem -accept 4433 -verify_return_error -www -state`. 

The `s_time` is specified here:
`openssl s_time -connect localhost:$4433 -CAfile $1_test/rootCA/demoCA/root.pem -new`

Before testing with `s_time`, I usually run `s_client` first to make sure that the handshake has no error. Here is my `s_client` command:
`openssl s_client -connect localhost:4433 -CAfile $1_test/rootCA/demoCA/root.pem -verify_return_error -state `

**Comment by judowha** [comment]
I can also provide files for another test described here [#363](https://github.com/open-quantum-safe/openssl/issues/363)

[test2.zip](https://github.com/openssl/openssl/files/8376239/test2.zip)

This test contains two folder `sp256_crsa_test` and `srsa_cp256_test`. 
`sp256_crsa_test` contains a root CA, a server and a client. The **server** has certificate signed by the root CA using **ECDSAP256**. The **client** has certificate signed by the root CA using **RSA3072**.
`srsa_cp256_test` contains a root CA, a server and a client. The **server** has certificate signed by the root CA using **RSA3072**. The **client** has certificate signed by the root CA using **ECDSAP256**.

Test commands I used are here:

Set up server
`openssl s_server -key $1/server/server.key -cert $1_test/server/server.pem -accept 4433 -CAfile $1/rootCA/demoCA/root.pem -verify_return_error -Verify 1 -www -state`

Test handshake 
`openssl s_client -key $1/client/client.key  -cert $1/client/client.pem -connect localhost:4433 -CAfile $1/rootCA/demoCA/root.pem -verify_return_error -state `

Test handshake performance
`openssl s_time -key $1/client/client.key -cert $1/client/client.pem -connect localhost:4433 -CAfile $1/rootCA/demoCA/root.pem -new`

**Comment by baentsch** [comment]
Thanks for sharing your setup. Below some observations:

> In theory, ECDSA should perform better than RSA in the same security level but the test result shows the opposite.

As per your measurements: 19755 P256 connections in 31 secs vs. 7380 RSA connections in 31 secs indicates that P256 is clearly faster than RSA -- as expected.

I do agree that the measurement statement about "user time" is pretty misleading, though. It surely is worth while understanding the deviation between wall clock time and "user time" collected by https://github.com/openssl/openssl/blob/db44b55aaa42141921217183667800425227b658/apps/s_time.c#L311

> I find that the test result for adding the -new flag and adding the -reuse flag is basically the same.

In my own runs, these measurements are not the same. They are similar, agreed, but consistently `-new` performs worse than `-renew` -- as one would expect it. I do agree that the difference (ca. 5%) is _very_ low, though: Shouldn't connection reuse be _much_ faster than new connection establishment? What puzzles me is that the letter 'r' is never output when running the connection reuse test as it would be expected when reading the code for this: https://github.com/openssl/openssl/blob/db44b55aaa42141921217183667800425227b658/apps/s_time.c#L377-L378

So you see the same ("*") output? If so, the question is: Why is connection reuse apparently never done even though the same `scon` object is passed at each iteration?  https://github.com/openssl/openssl/blob/db44b55aaa42141921217183667800425227b658/apps/s_time.c#L361

Some code digging required, it seems...

**Comment by baentsch** [comment]
> Is this resumption supported in OpenSSL?

I'd absolutely say so: See for example https://www.openssl.org/docs/manmaster/man3/SSL_psk_server_cb_func.html . What puzzles me though, is that `s_time` doesn't call [SSL_CTX_set_psk_use_session_callback](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_psk_use_session_callback.html). This to me would indicate that `s_time` doesn't handle PSK and accordingly doesn't "re-create" "session resume" in "TLS1.3 style". The one statement hinting at that is 

```
Because this program does not have all the options of the [openssl-s_client(1)](https://www.openssl.org/docs/man3.0/man1/openssl-s_client.html) program to turn protocols on and off, you may not be able to measure the performance of all protocols with all servers.
```
 (from https://www.openssl.org/docs/man3.0/man1/openssl-s_time.html)

--> TLS1.3 PSK seems to be such option (missing from `s_time` and hence not usable there). So I guess one would have to add the code for setting up PSK (and users also should not forget to set it when starting the server). 

This question in my eyes is resolved, thus.

It might lead to someone opening an explicit ticket to support TLS1.3 PSK to enable "TLS1.3 reuse" `s_time` runs, though.

> Then I found the code of double app_tminterval(int stop, int usertime)

This is platform-dependent code to determine user time that looks good to me. The real question is why is there such a big difference between user time and wall clock time -- and why does the difference seem larger for P256 than RSA?

The only one reason I can think of is that much more data is transported for RSA keys. As this time most likely is not accounted for in user time tracking, EC appears worse than RSA: More EC connections are made, requiring more user CPU time to process them. But then again, one could argue this reporting makes sense: only the time spent to _operate_ the EC/RSA crypto is counted -- which in turn seems to indicate better optimizations for RSA code to exist than for EC code. But that's something `openssl speed` should answer: What's the (3kRSA vs P256EC comparative) results there on your machine? At least on mine, P256 ECDSA verify is slower that 3kRSA verify -- which arguably explains the (lack in) compute time difference (per iteration).

In other words, these numbers (ECC/RSA [non-]difference) seem to make sense, also resolving this question.

**Comment by judowha** [comment]
Thanks for your patient reply, but I still have the last question. 

> What's the (3kRSA vs P256EC comparative) results there on your machine?

On my machine, 3kRSA indeed verifies faster than P256EC. I understand what you mean. 3kRSA has a larger size for key and signature so it needs more time to be processed. This processing time may not be included in user time tracking. This explanation indeed makes sense for this case.

However, this couldn't explain the second test I mentioned here: https://github.com/openssl/openssl/issues/17964#issuecomment-1082548884

Allow me to describe the test again:
I set two Groups of server-client:
- Group1 contains a Server1 with a RSA3072 certificate and a Client1 with a ECDSAP256 certificate
- Group2 contains a Server2 with a ECDSAP256 certificate and a Client2 with a RSA3072 certificate

Then I run mutual authentication test (both server and client need to check each other's certificate) using `s_time`. The test result is shown here: 
- Group1:
5696 connections in **5.46s**; **1043.22** connections/user sec, bytes read 0
5696 connections in 31 real seconds, 0 bytes read per connection

- Group2:
5652 connections in **23.03s**; **245.42** connections/user sec, bytes read 0
5652 connections in 31 real seconds, 0 bytes read per connection

In your theory, the user time and the speed for process handshake of these two groups should be roughly the same. But here the first group is significantly faster than the second group. 
Could you please help me explain this? Is it because the user time here only counts client verify server but does not count server verify client?

**Comment by baentsch** [comment]
> In your theory, the user time and the speed for process handshake of these two groups should be roughly the same. 

No, they should not be roughly the same: What the two setups/groups do is a) introduce private key operations on the client side and b) turn around the amount of public/private key processing on the client side: In Group 1 the client needs to do P256 signing, in Group 2 it needs to do RSA signing. Now, at least on my machine, RSA3k private key ops are much slower than P256 private key ops. Accordingly, Group 2 needs more time per connection. This measurement actually confirms my theory.

> Is it because the user time here only counts client verify server but does not count server verify client?

No, it's because of client auth effort/time accruing on the client side.

**Comment by judowha** [comment]
> No, they should not be roughly the same

As you said if we only consider the operation on the client-side, then the result is reasonable. However, I think the handshake should also contain the server-side. 
For group 1 the server needs to do RSA signing. For group 2 the server needs to do ECDSA signing. Therefore, for both groups, the handshake should contain one RSA signing, one ECDSA signing, two RSA verifying, and two ECDSA verifying. That is why I said the connection speed for these two groups should be roughly the same. 

> No, it's because of client auth effort/time accruing on the client-side.

Sorry that I am not very clear about what does this sentence mean. Do you mean that `s_time` only measures the auth time on the client side? 

If you mean that `s_time` only measures the user auth time on the client-side then indeed group1 should be faster than group2.

**Comment by romen** [comment]
> > No, they should not be roughly the same
> 
> As you said if we only consider the operation on the client-side, then the result is reasonable. However, I think the handshake should also contain the server-side. For group 1 the server needs to do RSA signing. For group 2 the server needs to do ECDSA signing. Therefore, for both groups, the handshake should contain one RSA signing, one ECDSA signing, two RSA verifying, and two ECDSA verifying. That is why I said the connection speed for these two groups should be roughly the same.
> 

They are roughly the same if you look at the number of total connections in the same amount of wall-clock time (31 secs).

For historical reasons, the average is reported in terms of connections / user-seconds and that is where you see a difference in the 2 groups: user-time is the time spent in userland (so it excludes all time spent between handing over data to the operating system to ship over TCP to the other peer, time the data travels over the network, time spent by the peer's OS handling received traffic, time spent on userland computation on the peer, time spent by the peer's OS to handle the response, time spent by the response to travel over the network, and finally time spent by the local OS to handle the incoming traffic and give it back to `s_time` userland to do some more computation).

When looking at user-time like `s_time` does by default, you have a client centric measure of how many handshakes per second you can do, so group1 and group2 looks drastically different in user-time because it changes which side is doing the more expensive operation!

**Comment by judowha** [comment]
OK, I now understand. After clarifying the definition of user time all the test results are now reasonable. Thank @baentsch and @romen for your warm reply! The question is solved so I will just close this problem. Again, truly appreciate your answer and your help!
